File: scripts/json_to_shopify_csv.py
import argparse
import csv
import json
import os
import re
import unicodedata
from typing import Any, Dict, Iterable, List, Optional, Tuple

SHOPIFY_HEADERS: List[str] = [
    "Handle",
    "Title",
    "Body (HTML)",
    "Vendor",
    "Type",
    "Tags",
    "Published",
    "Option1 Name",
    "Option1 Value",
    "Variant SKU",
    "Variant Grams",
    "Variant Inventory Tracker",
    "Variant Inventory Qty",
    "Variant Price",
    "Image Src",
]


def slugify(value: str) -> str:
    """Make a Shopify-safe handle (lowercase, hyphenated, ascii when possible)."""
    value = (value or "").strip().lower()
    value = unicodedata.normalize("NFKD", value)
    value = value.encode("ascii", "ignore").decode("ascii")
    value = re.sub(r"[^a-z0-9]+", "-", value)
    value = re.sub(r"-+", "-", value).strip("-")
    return value or "item"


def ensure_unique_handle(base: str, used: Dict[str, int]) -> str:
    if base not in used:
        used[base] = 1
        return base
    used[base] += 1
    return f"{base}-{used[base]}"


def parse_price(value: Any) -> str:
    # Shopify expects '.' decimal and no currency symbol.
    if value is None:
        return ""
    if isinstance(value, (int, float)):
        return f"{value:.2f}"
    s = str(value).strip()
    s = s.replace(",", ".")
    s = re.sub(r"[^0-9.]", "", s)
    if s == "":
        return ""
    try:
        return f"{float(s):.2f}"
    except ValueError:
        return ""


def as_tags(*parts: str) -> str:
    clean = [p.strip() for p in parts if (p or "").strip()]
    # Shopify tags are comma-separated.
    # Deduplicate while preserving order.
    seen = set()
    out: List[str] = []
    for p in clean:
        key = p.lower()
        if key in seen:
            continue
        seen.add(key)
        out.append(p)
    return ", ".join(out)


def resolve_image_src(image_path: Optional[str], public_base_url: str = "") -> str:
    if not image_path:
        return ""
    image_path = image_path.strip()
    if image_path.startswith("http://") or image_path.startswith("https://"):
        return image_path
    if public_base_url:
        return public_base_url.rstrip("/") + "/" + image_path.lstrip("/")
    # Shopify imports do best with absolute URLs, but we allow relative for local workflows.
    return image_path


def iter_menu_items(menu_json: Dict[str, Any]) -> Iterable[Tuple[Dict[str, Any], Dict[str, Any]]]:
    """Yield (category, item) pairs from src/data/menu.json."""
    for cat in menu_json.get("categories", []) or []:
        items = cat.get("items", []) or []
        for item in items:
            yield cat, item


def convert_menu_to_rows(
    menu_json: Dict[str, Any],
    vendor: str,
    published: str,
    inventory_tracker: str,
    inventory_qty: str,
    grams: str,
    public_base_url: str,
) -> List[List[str]]:
    used_handles: Dict[str, int] = {}
    rows: List[List[str]] = []

    for cat, item in iter_menu_items(menu_json):
        title = str(item.get("title") or "").strip()
        if not title:
            continue

        category_name = str(cat.get("name") or "").strip()
        category_slug = str(cat.get("slug") or "").strip()

        amount = str(item.get("amount") or "").strip()
        option1_name = "Size" if amount else "Title"
        option1_value = amount if amount else "Default Title"

        body_text = str(item.get("text") or "").strip()
        # If no long description, use category description as fallback.
        if not body_text:
            body_text = str(cat.get("description") or "").strip()
        body_html = body_text

        price = parse_price(item.get("price"))

        # Handle: base on title + amount to reduce collisions.
        base_handle = slugify(title if not amount else f"{title} {amount}")
        handle = ensure_unique_handle(base_handle, used_handles)

        sku = slugify(handle).upper()

        tags = as_tags(category_name, category_slug)
        img_src = resolve_image_src(item.get("image"), public_base_url=public_base_url)

        rows.append(
            [
                handle,
                title,
                body_html,
                vendor,
                category_name or "Menu",
                tags,
                published,
                option1_name,
                option1_value,
                sku,
                grams,
                inventory_tracker,
                inventory_qty,
                price,
                img_src,
            ]
        )

    return rows


def convert_products_json_to_rows(
    products_json: Dict[str, Any],
    vendor: str,
    published: str,
    inventory_tracker: str,
    inventory_qty: str,
    grams: str,
    public_base_url: str,
) -> List[List[str]]:
    used_handles: Dict[str, int] = {}
    rows: List[List[str]] = []

    items = products_json.get("items", []) or []
    for item in items:
        title = str(item.get("title") or "").strip()
        if not title:
            continue

        body_html = str(item.get("short") or "").strip()
        price = parse_price(item.get("price"))
        image = resolve_image_src(item.get("image"), public_base_url=public_base_url)

        base_handle = slugify(title)
        handle = ensure_unique_handle(base_handle, used_handles)
        sku = slugify(handle).upper()

        tags = as_tags("products")

        rows.append(
            [
                handle,
                title,
                body_html,
                vendor,
                "Product",
                tags,
                published,
                "Title",
                "Default Title",
                sku,
                grams,
                inventory_tracker,
                inventory_qty,
                price,
                image,
            ]
        )

    return rows


def main() -> int:
    parser = argparse.ArgumentParser(description="Convert deliorman JSON to Shopify product CSV")
    parser.add_argument(
        "--input",
        default=os.path.join("src", "data", "menu.json"),
        help="Path to input JSON (default: src/data/menu.json)",
    )
    parser.add_argument(
        "--kind",
        choices=["menu", "products"],
        default="menu",
        help="Which JSON shape to parse (default: menu)",
    )
    parser.add_argument(
        "--output",
        default="shopify_import.csv",
        help="Output CSV filename (default: shopify_import.csv)",
    )
    parser.add_argument("--vendor", default="Deliorman", help="Shopify Vendor field")
    parser.add_argument("--published", default="TRUE", help="Published TRUE/FALSE")
    parser.add_argument(
        "--inventory-tracker",
        default="",
        help="Variant Inventory Tracker (blank, 'shopify', etc.)",
    )
    parser.add_argument(
        "--inventory-qty",
        default="0",
        help="Variant Inventory Qty (default: 0)",
    )
    parser.add_argument(
        "--grams",
        default="0",
        help="Variant Grams (default: 0)",
    )
    parser.add_argument(
        "--public-base-url",
        default="",
        help="If set, relative image paths become absolute URLs (e.g. https://example.com)",
    )

    args = parser.parse_args()

    with open(args.input, "r", encoding="utf-8") as f:
        data = json.load(f)

    if args.kind == "menu":
        rows = convert_menu_to_rows(
            data,
            vendor=args.vendor,
            published=args.published,
            inventory_tracker=args.inventory_tracker,
            inventory_qty=str(args.inventory_qty),
            grams=str(args.grams),
            public_base_url=args.public_base_url,
        )
    else:
        rows = convert_products_json_to_rows(
            data,
            vendor=args.vendor,
            published=args.published,
            inventory_tracker=args.inventory_tracker,
            inventory_qty=str(args.inventory_qty),
            grams=str(args.grams),
            public_base_url=args.public_base_url,
        )

    with open(args.output, "w", newline="", encoding="utf-8") as f:
        writer = csv.writer(f)
        writer.writerow(SHOPIFY_HEADERS)
        writer.writerows(rows)

    print(f"Wrote {len(rows)} product rows to {args.output}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/json_to_shopify_csv.mjs
import fs from 'node:fs';
import path from 'node:path';

const SHOPIFY_HEADERS = [
  'Handle',
  'Title',
  'Body (HTML)',
  'Vendor',
  'Type',
  'Tags',
  'Published',
  'Option1 Name',
  'Option1 Value',
  'Variant SKU',
  'Variant Grams',
  'Variant Inventory Tracker',
  'Variant Inventory Qty',
  'Variant Price',
  'Image Src',
];

function csvEscape(value) {
  const s = String(value ?? '');
  if (/[\r\n",]/.test(s)) return `"${s.replaceAll('"', '""')}"`;
  return s;
}

function writeCsv(filePath, rows) {
  const lines = [];
  lines.push(SHOPIFY_HEADERS.map(csvEscape).join(','));
  for (const row of rows) {
    lines.push(row.map(csvEscape).join(','));
  }
  fs.writeFileSync(filePath, lines.join('\r\n') + '\r\n', 'utf8');
}

function normalizeTitle(s) {
  return String(s ?? '')
    .trim()
    .replace(/\s+/g, ' ')
    .toLocaleLowerCase('bg-BG');
}

function slugifyBgToHandle(value) {
  // Shopify handles are typically ascii. We'll transliterate Bulgarian Cyrillic -> Latin.
  const s = String(value ?? '').trim().toLocaleLowerCase('bg-BG');
  if (!s) return 'item';

  const map = {
    'а': 'a',
    'б': 'b',
    'в': 'v',
    'г': 'g',
    'д': 'd',
    'е': 'e',
    'ж': 'zh',
    'з': 'z',
    'и': 'i',
    'й': 'y',
    'к': 'k',
    'л': 'l',
    'м': 'm',
    'н': 'n',
    'о': 'o',
    'п': 'p',
    'р': 'r',
    'с': 's',
    'т': 't',
    'у': 'u',
    'ф': 'f',
    'х': 'h',
    'ц': 'ts',
    'ч': 'ch',
    'ш': 'sh',
    'щ': 'sht',
    'ъ': 'a',
    'ь': '',
    'ю': 'yu',
    'я': 'ya',
  };

  let out = '';
  for (const ch of s) {
    out += map[ch] ?? ch;
  }

  out = out
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/-+/g, '-')
    .replace(/^-|-$/g, '');

  return out || 'item';
}

function ensureUniqueHandle(base, used) {
  const count = used.get(base) ?? 0;
  used.set(base, count + 1);
  return count === 0 ? base : `${base}-${count + 1}`;
}

function parsePrice(value) {
  if (value == null) return '';
  if (typeof value === 'number' && Number.isFinite(value)) return value.toFixed(2);
  const s = String(value).trim().replace(',', '.').replace(/[^0-9.]/g, '');
  if (!s) return '';
  const n = Number.parseFloat(s);
  return Number.isFinite(n) ? n.toFixed(2) : '';
}

function asTags(...parts) {
  const clean = parts.map((p) => String(p ?? '').trim()).filter(Boolean);
  const seen = new Set();
  const out = [];
  for (const p of clean) {
    const key = p.toLocaleLowerCase('bg-BG');
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(p);
  }
  return out.join(', ');
}

function resolveImageSrc(imagePath, publicBaseUrl) {
  const p = String(imagePath ?? '').trim();
  if (!p) return '';
  if (/^https?:\/\//i.test(p)) return p;
  if (publicBaseUrl) return publicBaseUrl.replace(/\/+$/, '') + '/' + p.replace(/^\/+/, '');
  return p; // allow relative
}

function convertMenuJsonToRows(menu, opts) {
  const usedHandles = new Map();
  const rows = [];

  for (const cat of menu.categories ?? []) {
    const categoryName = String(cat?.name ?? '').trim();
    const categorySlug = String(cat?.slug ?? '').trim();

    for (const item of cat.items ?? []) {
      const title = String(item?.title ?? '').trim();
      if (!title) continue;

      const amount = String(item?.amount ?? '').trim();
      const option1Name = amount ? 'Размер' : 'Title';
      const option1Value = amount || 'Default Title';

      let body = String(item?.text ?? '').trim();
      if (!body) body = String(cat?.description ?? '').trim();

      const price = parsePrice(item?.price);

      const baseHandle = slugifyBgToHandle(amount ? `${title} ${amount}` : title);
      const handle = ensureUniqueHandle(baseHandle, usedHandles);

      // Keep SKU ASCII (Shopify-friendly)
      const sku = handle.replace(/-/g, '_').toUpperCase();

      const tags = asTags(categoryName, categorySlug);
      const imageSrc = resolveImageSrc(item?.image, opts.publicBaseUrl);

      rows.push([
        handle,
        title,
        body,
        opts.vendor,
        categoryName || 'Меню',
        tags,
        opts.published,
        option1Name,
        option1Value,
        sku,
        String(opts.grams),
        String(opts.inventoryTracker ?? ''),
        String(opts.inventoryQty),
        price,
        imageSrc,
      ]);
    }
  }

  return rows;
}

function parseArgs(argv) {
  const args = {
    input: path.join('src', 'data', 'menu.json'),
    output: 'shopify_import.csv',
    vendor: 'Делиорман',
    published: 'TRUE',
    inventoryTracker: '',
    inventoryQty: '0',
    grams: '0',
    publicBaseUrl: '',
  };

  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--input') args.input = argv[++i];
    else if (a === '--output') args.output = argv[++i];
    else if (a === '--vendor') args.vendor = argv[++i];
    else if (a === '--published') args.published = argv[++i];
    else if (a === '--inventory-tracker') args.inventoryTracker = argv[++i];
    else if (a === '--inventory-qty') args.inventoryQty = argv[++i];
    else if (a === '--grams') args.grams = argv[++i];
    else if (a === '--public-base-url') args.publicBaseUrl = argv[++i];
    else if (a === '--help' || a === '-h') args.help = true;
  }

  return args;
}

function printHelp() {
  console.log(`Usage: node scripts/json_to_shopify_csv.mjs [options]

Bulgarian-only exporter: reads src/data/menu.json and writes a Shopify products CSV.

Options:
  --input <path>           Input JSON (default: src/data/menu.json)
  --output <path>          Output CSV (default: shopify_import.csv)
  --vendor <text>          Vendor field (default: Делиорман)
  --published TRUE|FALSE   Published (default: TRUE)
  --inventory-tracker <t>  Variant Inventory Tracker (default: blank)
  --inventory-qty <n>      Variant Inventory Qty (default: 0)
  --grams <n>              Variant Grams (default: 0)
  --public-base-url <url>  Prefix for relative image paths
`);
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  if (args.help) {
    printHelp();
    process.exit(0);
  }

  const repoRoot = path.resolve(process.cwd());
  const inputPath = path.isAbsolute(args.input) ? args.input : path.join(repoRoot, args.input);
  const outputPath = path.isAbsolute(args.output) ? args.output : path.join(repoRoot, args.output);

  const raw = fs.readFileSync(inputPath, 'utf8');
  const menu = JSON.parse(raw);

  const rows = convertMenuJsonToRows(menu, args);
  writeCsv(outputPath, rows);

  console.log(`Wrote ${rows.length} rows to ${path.relative(repoRoot, outputPath)}`);
}

main();

